### PR TITLE
[Spark] Add redis support

### PR DESCRIPTION
**Pull request by alxtkr77**
[ML-3008](https://jira.iguazeng.com/browse/ML-3008)

### CODE REVIEW

File: tests/system/feature_store/test_spark_engine.py [hunk 1]
```diff
@@ -292,6 +297,58 @@ def test_ingest_to_csv(self):
             read_back_df_storey.sort_index(axis=1)
         )
 
+    @pytest.mark.parametrize("spark_is_local", [False])
```
Comment (assaf758) [inline tests/system/feature_store/test_spark_engine.py:300]: no need to parametrize

File: tests/system/feature_store/test_spark_engine.py [hunk 2]
```diff
@@ -292,6 +297,58 @@ def test_ingest_to_csv(self):
...
+            entities=[fstore.Entity(key)],
+            timestamp_key="timestamp",
+            engine="spark",
+        )
+        source = ParquetSource("myparquet", path=self.get_remote_pq_source_path())
+        redis_path = mlrun.get_secret_or_env("MLRUN_REDIS_URL")
+        targets = [RedisNoSqlTarget(path=redis_path)]
+        measurements.set_targets(targets, with_defaults=False)
+        if spark_is_local:
+            from pyspark.sql import SparkSession
```
Comment (assaf758) [inline tests/system/feature_store/test_spark_engine.py:316]: please move to top of file

File: tests/system/feature_store/test_spark_engine.py [hunk 3]
```diff
@@ -292,6 +297,58 @@ def test_ingest_to_csv(self):
...
+        name = "measurements_spark"
+
+        measurements = fstore.FeatureSet(
+            name,
+            entities=[fstore.Entity(key)],
+            timestamp_key="timestamp",
+            engine="spark",
+        )
+        source = ParquetSource("myparquet", path=self.get_remote_pq_source_path())
+        redis_path = mlrun.get_secret_or_env("MLRUN_REDIS_URL")
```
Comment (assaf758) [inline tests/system/feature_store/test_spark_engine.py:312]: use MLRUN_REDIS__URL in env.yml

**Review by assaf758 — CHANGES_REQUESTED** [review]

File: mlrun/datastore/targets.py [hunk 4]
```diff
@@ -1176,6 +1176,33 @@ def get_table_object(self):
             flush_interval_secs=mlrun.mlconf.feature_store.flush_interval,
         )
 
+    def get_spark_options(self, key_column=None, timestamp_key=None, overwrite=True):
+        from urllib.parse import urlparse
```
Comment (assaf758) [inline mlrun/datastore/targets.py:1180]: move to start of file

File: mlrun/datastore/targets.py [hunk 5]
```diff
@@ -1176,6 +1176,33 @@ def get_table_object(self):
...
+            "table": "{" + store_path_to_spark(self.get_target_path()),
+            "format": "org.apache.spark.sql.redis",
+            "host": parsed_url.hostname,
+            "port": parsed_url.port if parsed_url.port else "6379",
+            "user": parsed_url.username,
+            "auth": parsed_url.password,
+        }
+
+    def prepare_spark_df(self, df, key_columns):
+        from pyspark.sql.functions import udf
```
Comment (assaf758) [inline mlrun/datastore/targets.py:1200]: move to start of file

File: mlrun/datastore/targets.py [hunk 6]
```diff
@@ -1176,6 +1176,33 @@ def get_table_object(self):
             flush_interval_secs=mlrun.mlconf.feature_store.flush_interval,
         )
 
+    def get_spark_options(self, key_column=None, timestamp_key=None, overwrite=True):
+        from urllib.parse import urlparse
+
+        parsed_url = urlparse(self.get_target_path())
+        if parsed_url.hostname is None:
+            self.path = self._secrets.get(
```
Comment (assaf758) [inline mlrun/datastore/targets.py:1184]: instead of env var, use mlrun.mlconf.redis.url

File: mlrun/datastore/utils.py [hunk 7]
```diff
@@ -16,7 +16,11 @@
 
 
 def store_path_to_spark(path):
-    if path.startswith("v3io:///"):
+    if path.startswith("redis://") or path.startswith("rediss://"):
+        url = urlparse(path)
+        if url.path:
+            return url.path
```
Comment (assaf758) [inline mlrun/datastore/utils.py:22]: suggest keeping with the function structure of setting "path" and returning it in the end of function.
also what is the handling if not url.path exist?

**Review by assaf758 — CHANGES_REQUESTED** [review]

File: mlrun/datastore/utils.py (hunk 7 above)
Comment (alxtkr77) [reply mlrun/datastore/utils.py:22]: Will move. Non existing url.path is handled by the caller

File: mlrun/datastore/targets.py (hunk 5 above)
Comment (alxtkr77) [reply mlrun/datastore/targets.py:1200]: ok

File: mlrun/datastore/targets.py (hunk 4 above)
Comment (alxtkr77) [reply mlrun/datastore/targets.py:1180]: ok

File: mlrun/datastore/targets.py [hunk 8]
```diff
@@ -1190,6 +1193,26 @@ def get_table_object(self):
             flush_interval_secs=mlrun.mlconf.feature_store.flush_interval,
         )
 
+    def get_spark_options(self, key_column=None, timestamp_key=None, overwrite=True):
+        parsed_url = urlparse(self.get_target_path())
+        if parsed_url.hostname is None:
+            self.path = mlrun.mlconf.redis.url
+            parsed_url = urlparse(self.get_target_path())
```
Comment (assaf758) [inline mlrun/datastore/targets.py:1200]: not  parsed_url = urlparse(self.path)?

Comment (alxtkr77) [reply mlrun/datastore/targets.py:1200]: In all the places we use ' urlparse(self.get_target_path())'. In this case it probably returns the self.path. But I wanted to keep it generic

**Review by assaf758 — APPROVED** [review]

**Pull request merged**